Methyl dragon/fixups (#1)
wjwwood/rep: rep-2010.rst

The proposed patterns use the existing default serialization technology, i.e. CDR, in combination with new tooling.
However, technologies provided by different, perhaps future, serialization technologies can always be used in addition.

> Specifically, this REP proposes that the following is a good way to achieve backwards compatibility in messages over long periods of time and in a variety of scenarios, e.g. over the wire, converting bag files, or in specialized tools:
> 
> - Interpreting older versions of the message at runtime;
> - Using its type information;
> - And converting it to the current version of the message using user defined transfer functions

Reviewer: ```suggestion
Specifically, this REP proposes that interpreting older versions of the message at runtime can be achieved by using its type information and then converting it to the current version of the message using user-defined transfer functions.
This approach is a good way to achieve backwards compatibility in messages over long periods of time and in a variety of scenarios, e.g. over the wire, converting bag files, or in specialized tools.
```

> 

Author: I don't really care for this change. It inverts the information presented, and it makes it more confusing in my opinion.

I think a better thing to do would be to break it into two sentences instead:

```
Specifically, this REP proposes that interpreting older versions of the message at runtime can be achieved by using its type information and then converting it to the current version of the message using user-defined transfer functions.
This approach is a good way to achieve backwards compatibility in messages over long periods of time and in a variety of scenarios, e.g. over the wire, converting bag files, or in specialized tools.
```

Author: Also, for future reference, I don't think putting `;` at the end of bullets points is needed. And unless they are titles or start with proper nouns, I wouldn't capitalize the first word in them either.

The "transfer function" can be user-defined, or for simple changes (like changing the field type to a compatible type) it can be done automatically.
We already do something like this for the ROS 1 to ROS 2 bridge in order to handle changes between message types in ROS 1 and ROS 2, and something like this was also done for rosbags in ROS 1 with the "bag migration rules" feature.

> Furthermore, the "transfer function" approach also allows for runtime transformation of messages that only change in field semantics (e.g., a change of ``float distance`` to mean centimeters instead of meters, but with no change in field name or type.) Although in those cases, users will likely have to define the transfer function themselves.

Author: ```suggestion
Furthermore, the "transfer function" approach also allows for conversions between messages which are only semantically different (e.g., a change in semantics for a field ``float distance`` to mean centimeters instead of meters, but with no change in field name or type).
However, the user would have to write the transfer function for these changes, because it could not be made automatically as they are not captured in a structured way.
```

.. TODO:: cite the above

This approach requires a few features, like the ability to have a single application read old and new versions of a message at the same time, and it requires more infrastructure and tooling to make it work, but it has the advantage of keeping both the publishing and subscribing code simple, i.e agnostic to the fact that there are other versions of the message, and it keeps the message type from being cluttered with vestigial fields.

Either way, a problem can usually be solved by changing a message in some of, if not all, of the of the above mentioned ways, and is often influenced by what the underlying technology allows for or encourages.
ROS 2 has special considerations on this topic because it can support different serialization technologies, though CDR is the default and most common right now, and those technologies have different capabilities.

> 

Author: ```suggestion
```

Type Version Hash
~~~~~~~~~~~~~~~~~

The type version hashes are not sequential and do not imply any rank among versions of the type. That is, given two version hashes of a type, there is no way to tell which is "newer".

> The type version hash can only be used to determine if type versions are equal and if there exists a chain of transfer functions that can convert between them.

Author: 👍 good catch

Because of this, when a change to a type is made, it may or may not be necessary to write transfer functions in both directions depending on how the interface is used.

In order to calculate the type version hashes so that they are stable and are not sensitive to trivial changes like changes in the comments or whitespace in the IDL file, the IDL file given by the user, which may be a ``.msg`` file, ``.idl`` file, or something else, is parsed and stored into a data structure which excludes things like comments but includes things that impact compatibility on the wire.

The data structure includes:

- A list of field names and types, but not default values
- The serialization format
- The serialization format version

> - An optional user-defined interface version, or 0 if not provided

Author: Let's not capitalize these.

> - A list of field names and types, but not default values
> - The serialization format
> - The serialization format version
> - An optional user-defined interface version, or 0 if not provided

Reviewer: ```suggestion
- a list of field names and types, but not default values
- the serialization format
- the serialization format version
- an optional user-defined interface version, or 0 if not provided
```

The resulting data structure is hashed using a standard SHA-1 method, resulting in a standard 160-bit (20-byte) hash value which is also generally known as a "message digest".
This hash is combined with a "type version hash standard version", the first of which will be ``IDLHASH-1``, with an ``@`` symbol, resulting in a complete type version hash like ``IDLHASH-1@<160-bit SHA-1 of data structure>``.
This allows the tooling to know if a hash mismatch is due to a change in this standard (what is being hashed) or due to a difference in the interface types themselves.

> Notably, the user-defined interface version being included in the hash allows for messages that only change in field semantics (i.e., without changing field names or types) to be picked up as discrepancies when they would not before, allowing users to be prompted to write "transfer functions" to resolve them.

Author: ```suggestion
The optional user-defined interface version included in the hash allows the user to indicate that there was a change in message semantics (i.e. no change in the field names or types), causing the versions to mismatch and allowing the user to write a "transfer functions" to resolve the semantic differences.
```

.. TODO:: is the list of field names and types sufficient? how to capture things like .idl annotations, etc... I'm thinking of serialization format specific entries can be added to this data structure, but need to sketch it out a bit more

Enforcing Type Version
~~~~~~~~~~~~~~~~~~~~~~

> If the type version hash is available, it can be used as an additional constraint to determine if two endpoints (publishers and subscriptions) on a topic should communicate.

Author: ```suggestion
The type version hash can be used as an additional constraint to determine if two endpoints (publishers and subscriptions) on a topic should communicate.
```

When creating a publisher or subscription, the caller normally provides: a topic name, QoS settings, and a topic type.
The topic type is represented as a string and is automatically deduced based on the type given to the create function, e.g. as a template parameter in C++ or the message type as an argument in Python.
For example, creating a publisher for ``std_msgs::msg::String`` in C++, may result in a topic type like ``std_msgs/msg/String``.
All of these items are used by the middleware to determine if two endpoints should communicate or not, and this REP proposes that the type version be added to this list of provided information.

> From the user's perspective, nothing needs to change, as the type version can be extracted based on the topic type given either at the ``rcl`` layer or in the ``rmw`` implementation itself.

Author: ```suggestion
Nothing needs to change from the user's perspective, as the type version can be extracted automatically based on the topic type given, either at the ``rcl`` layer or in the ``rmw`` implementation itself.
```
